### DISCUSSION TITLE
[XMLPARSER Error]  The profile's envirment can't parse to interger

### DISCUSSION THREAD

**Discussion by ken20020209** [eProsima/Fast-DDS#4577] [Support]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

The profile \<port>${port]\<port> will get the env parameter. Then the app can run at different ports.

### Current behavior

The error occurred. The Function getXMLUint receives the wrong type parameter.

### Steps to reproduce

export FASTRTPS_DEFAULT_PROFILES_FILE=super_client_configuration_file.xml
export DISCOVERY_SERVER_PORT=11811
ros2 run topic list --no-daemon

### Fast DDS version/commit

latest version:725f1c7
and
foxy original version.

### Platform/Architecture

Ubuntu Focal 20.04 amd64

### Transport layer

UDPv4

### Additional context

ros2 foxy version

the enviroment setting ${var} can run sucess in \<address>${ip}\</address>

### XML configuration file

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<dds>
    <profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
        <participant profile_name="super_client_profile" is_default_profile="true">
            <rtps>
                <builtin>
                    <discovery_config>
                        <discoveryProtocol>SUPER_CLIENT</discoveryProtocol>
                        <discoveryServersList>
                            <RemoteServer prefix="44.53.00.5f.45.50.52.4f.53.49.4d.41">
                                <metatrafficUnicastLocatorList>
                                    <locator>
                                        <udpv4>
                                            <address>127.0.0.1</address>
                                            <port>${DISCOVERY_SERVER_PORT}</port>
                                        </udpv4>
                                    </locator>
                                </metatrafficUnicastLocatorList>
                            </RemoteServer>
                        </discoveryServersList>
                    </discovery_config>
                </builtin>
            </rtps>
        </participant>
    </profiles>
</dds>
```

### Relevant log output

```shell
2024-03-18 00:15:34.077 [XMLPARSER Error] <port> getXMLUint XML_ERROR! -> Function getXMLUint
2024-03-18 00:15:34.077 [XMLPARSER Error] getXMLRemoteServer was given a misformatted server's metatrafficUnicastLocatorList XML_ERROR! -> Function getXMLRemoteServer
2024-03-18 00:15:34.077 [XMLPARSER Error] getXMLList was given a misformatted RemoteServer XML_ERROR! -> Function getXMLList
2024-03-18 00:15:34.077 [XMLPARSER Error] Error parsing participant profile -> Function parseXMLParticipantProf
2024-03-18 00:15:34.077 [XMLPARSER Error] Error parsing profile's tag participant -> Function parseProfiles
2024-03-18 00:15:34.077 [XMLPARSER Error] Error parsing 'super_client_configuration_file.xml' -> Function loadXMLFile
```

**Comment by JesusPoderoso** [comment]
Hi @ken20020209, thanks for using Fast DDS.
[Processing environment variables in XML text](https://fast-dds.docs.eprosima.com/en/v2.12.0/fastdds/xml_configuration/making_xml_profiles.html#xml-environment-variables) feature was introduced in [Fast DDS v2.12.0](https://fast-dds.docs.eprosima.com/en/v2.12.0/notes/notes.html). 

As long as ROS 2 Foxy is EOL, as well as Fast DDS v2.1.x (Fast DDS version attached to the distro), that feature was not included in your ROS distribution. For that reason, I am moving this ticket to the corresponding forum as per [Fast DDS CONTRIBUTING guidelines](https://github.com/eProsima/Fast-DDS/blob/master/CONTRIBUTING.md).

It is going to be difficult if you want to use that feature.  On  the one hand, it is highly difficult to update the Fast DDS version from v2.1.x up to v2.12.x / latest Fast DDS version. A lot of features and new API have been included or changed.
I would recommend you update your ROS 2 distribution up to the latest (ROS 2 iron), which contains Fast DDS version v2.10.x, and then update the Fast DDS version up to v2.12.x.
On the other hand, another solution would be waiting a few months until [ROS 2 Jazzy](https://docs.ros.org/en/rolling/The-ROS2-Project/Roadmap.html#id2) is released, which would contain the latest Fast DDS version (which would include the feature).

Hope it helps, and sorry for the inconvenience!

**Comment by JesusPoderoso** [comment]
There is a third option, which is to migrate to eProsima's custom ROS 2 distribution, [Vulcanexus](https://vulcanexus.org/). It is the all-in-one ROS 2 toolset that contains the latest versions of ALL eProsima products, in the same distributions as ROS 2 (humble, iron...).
Vulcanexus always includes the latest Fast DDS version, no matter which ROS 2 distribution you choose.
